File: java/src/main/java/com/fleetmgr/sdk/client/state/pilot/Disconnecting.java
package com.fleetmgr.sdk.client.state.pilot;

import com.fleetmgr.sdk.client.event.input.connection.ConnectionEvent;
import com.fleetmgr.sdk.client.event.input.user.UserEvent;
import com.fleetmgr.sdk.client.event.output.facade.FacadeEvent;
import com.fleetmgr.sdk.system.machine.BaseState;

import java.util.Optional;

/**
 * Created by: Bartosz Nawrot
 * Date: 23.09.2018
 * Description:
 */
public class Disconnecting extends BaseState {

    private boolean wasRecovering;

    Disconnecting(BaseState state, boolean wasRecovering) {
        super(state);
        this.wasRecovering = wasRecovering;
    }

    @Override
    public void start() {
        backend.getHeartbeatHandler().end();

        if (wasRecovering) {
            backend.closeFacadeChannel();
            listener.onEvent(new FacadeEvent(FacadeEvent.Type.RELEASED));
        }
    }

    @Override
    public Optional<BaseState> getInitialState() {
        if (wasRecovering) {
            return Optional.of(new Disconnected(this));
        } else {
            return Optional.empty();
        }
    }

    @Override
    public Optional<BaseState> onUserEvent(UserEvent event) {
        switch (event.getType()) {
            default:
                return defaultConnectionEventHandler(event.toString());
        }
    }

    @Override
    public Optional<BaseState> onConnectionEvent(ConnectionEvent event) {
        switch (event.getType()) {
            case CLOSED:
                backend.closeFacadeChannel();
                listener.onEvent(new FacadeEvent(FacadeEvent.Type.RELEASED));
                return Optional.of(new Disconnected(this));

            default:
                return defaultConnectionEventHandler(event.toString());
        }
    }

    @Override
    public String toString() {
        return "Disconnecting";
    }
}

File: cpp/source/state/device/Disconnecting.cpp
#include "state/device/Disconnecting.hpp"

#include "state/device/Disconnected.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::device;

using namespace com::fleetmgr::interfaces;
using namespace com::fleetmgr::interfaces::facade::control;

using namespace event::input::user;
using namespace event::input::connection;
using namespace event::output;

Disconnecting::Disconnecting(IState& state) :
    IState(state)
{
}

std::unique_ptr<IState> Disconnecting::start()
{
    backend.getHeartbeatHandler().end();
    backend.closeFacadeConnection();
    listener.onEvent(std::make_shared<FacadeEvent>(FacadeEvent::RELEASED));
    return std::make_unique<Disconnected>(*this);
}

std::unique_ptr<IState> Disconnecting::handleUserEvent(const UserEvent& event)
{
    return defaultEventHandle(event.toString());
}

std::unique_ptr<IState> Disconnecting::handleConnectionEvent(const ConnectionEvent& event)
{
    return defaultEventHandle(event.toString());
}

std::string Disconnecting::toString() const
{
    return "Disconnecting";
}
